### PR TITLE
Issue/5720 simple payments taxes

### LINKED ISSUE #5720: Simple payments: Show tax rate from the backend
When "Charge taxes" is enabled in simple payments but the store doesn't charge taxes, instead of hiding the tax rate show "Tax (0.0%)" as on iOS. If the store does charge taxes, we need to show the tax rate based on the backend rather than calculate it in the client.

Both of these will require [some FluxC work](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2252) to add `tax_lines` to the order model.

### PR DESCRIPTION
**Pull request by nbradbury**
Closes #5720 - When "Charge taxes" is enabled in simple payments but the store doesn't charge taxes, instead of hiding the tax rate it now shows "Tax (0.00%)" as on iOS. If the store does charge taxes, we show the tax rate based on the backend rather than calculate it in the client.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt [hunk 1]
```diff
@@ -189,6 +191,15 @@ data class Order(
         val totalTax: BigDecimal,
     ) : Parcelable
 
+    @Parcelize
+    data class TaxLine(
+        val id: Long,
+        val compound: Boolean,
+        val taxTotal: String,
+        val shippingTaxTotal: String,
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt]: ❓ I don't have the domain knowledge so take this question with a grain of salt: Do we perhaps also need to take into account `feeTaxes` or are those unrelated?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragment.kt [hunk 2]
```diff
@@ -101,6 +101,8 @@ class SimplePaymentsFragment : BaseFragment(R.layout.fragment_simple_payments) {
                 if (chargeTaxes) {
                     binding.containerTaxes.isVisible = true
                     binding.textTaxMessage.isVisible = true
+                    binding.textTaxLabel.text =
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragment.kt:104]: 💡 We might want to consider adding unit tests for this, since it's technically a business logic. Wdyt?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragmentViewModel.kt [hunk 3]
```diff
@@ -33,6 +33,14 @@ class SimplePaymentsFragmentViewModel @Inject constructor(
             customerNote = viewState.customerNote
         )
 
+    // returns the tax rate as a string from the first tax line item if it exists
+    val taxRatePercent
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragmentViewModel.kt:41]: 💡 Same as above, we might want to consider adding unit tests for this.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragmentViewModel.kt [hunk 4]
```diff
@@ -33,6 +33,14 @@ class SimplePaymentsFragmentViewModel @Inject constructor(
...
+        get() = if (order.taxLines.isNotEmpty()) {
+            order.taxLines[0].ratePercent.toString()
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragmentViewModel.kt:43]: - ❓ Is it safe to take into account only the first element of the array/list? 
- 💡 We might want to consider logging an error when there are multiple items in the list - assuming it's not expected (I don't have the domain knowledge). 
- 💡 A comment with explanation why it's okay to ignore the other elements might be beneficial for future as well, wdyt?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragmentViewModel.kt [hunk 5]
```diff
@@ -33,6 +33,14 @@ class SimplePaymentsFragmentViewModel @Inject constructor(
             customerNote = viewState.customerNote
         )
 
+    // returns the tax rate as a string from the first tax line item if it exists
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragmentViewModel.kt]: Nitpick 🔍 : I'm not sure this comment is useful, since it merely describes what the very straightforward code does. Moreover, it's already misleading since it doesn't say what happens when the `line item` doesn't exist. Imo in the long run, the comment might get even more misleading when we update the code and forget to update the comment. Wdyt?

**Review by malinajirka — COMMENTED** [review]
Thanks for working on this @nbradbury! I don't have the domain knowledge so I asked a couple newbie questions. The code looks good overall, when we add unit tests I think it'll be ready for merge.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt (hunk 1 above)
Comment (nbradbury) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt]: @malinajirka  I've not heard of `feeTaxes` and I don't see them anywhere in our code. What are they?

Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt]: Sorry, I should have been more explicit. 

My understanding is that Order can contain different items
- [products](https://github.com/woocommerce/woocommerce-android/blob/9abd919c51ae2cd54f32c2d4a8291a3c4d11b0c5/WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt#L79)
- [shipping items](https://github.com/woocommerce/woocommerce-android/blob/9abd919c51ae2cd54f32c2d4a8291a3c4d11b0c5/WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt#L176)
- [fee items](https://github.com/woocommerce/woocommerce-android/blob/9abd919c51ae2cd54f32c2d4a8291a3c4d11b0c5/WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt#L185)

It seems each of these contains `totalTax` field. 

This PR mentions `taxTotal` and `shippingTaxTotal`, so I was just wondering if it should also include `feeTaxTotal`. I'm really more of just guessing here, I don't have anything to back this up - I simply noticed that shipping is mentioned but fees are not. Wdyt?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragment.kt (hunk 2 above)
Comment (nbradbury) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragment.kt:104]: @malinajirka I'm a little confused over what exactly I should test here and below. Could you elaborate?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt (hunk 1 above)
Comment (nbradbury) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt]: Ah, gotcha! The shipping tax is actually unused by the app, but I added it to the model since it's available from FluxC. But I'll simply remove it from the WCAndroid model to avoid confusion, and if it's needed down the road it's easy enough to add it back.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragment.kt (hunk 2 above)
Comment (nbradbury) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragment.kt:104]: @malinajirka I added the tests I _think_ you're suggesting in [SimplePaymentsFragmentViewModelTests.kt](https://github.com/woocommerce/woocommerce-android/blob/a2570b98cfa0b068cba4dfa2d7c27fa540c9607e/WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragmentViewModelTests.kt). If you're suggesting something different, please let me know.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt (hunk 1 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt]: >  The shipping tax is actually unused by the app

I see, that's what I was missing, thanks!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragment.kt (hunk 2 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragment.kt:104]: @nbradbury Tbh I missed that this change is in Fragment and not the ViewModel when I was writing the above comment.

The tests you wrote are perfect and exactly what I had in mind for [the other comment](https://github.com/woocommerce/woocommerce-android/pull/5731/files/fbaad44b24e7b9033edb95eba616905c8e0efc9c#r795537511). Here I was thinking we should check that the ViewState contain the correct string with the correct parameter. However, since this is in the fragment not the ViewModel we can't write a test for it.

**I wouldn't worry about it in this PR**, but I'd still like to share my opinion about the approach used in this Fragment+ViewModel. As mentioned above, it's not possible to write unit tests which would verify that the correct string resource with correct parameters is used. 

Imo ideally, the ViewModel would provide a ViewState which would already contain all the string resources + parameters + information about visibility of all the views. 

The current approach puts business logic into the view layer -> eg. the view layer needs to know that it should use `R.string.simple_payments_tax_with_percent` for `textTaxLabel` and that the parameter it should pass is `viewModel.taxRatePercent`.

It also needs to know that when `chargeTaxes` is `true` containerTaxes and `textTaxMessage` should be visible. And again, we can't write unit tests for it.

To give a concrete example, I think the ViewState could look something like this for the screen:
```
data class ViewState(
    val subtotal : UiString [it should already contain formatted currency)
    val totalTax : UiString [it should already contain formatted currency)
    val total : UiString [it should already contain formatted currency)
    val textTaxMessage: UiString? [When it's null `textTaxMessage` view should not be visible - we can use `UiHelpers` for that]
    val chargeTaxesChecked: Boolean
   val customerNote: UiString
    ) : Parcelable
```

Also, all the navigation which currently lives in the fragment (eg. `showCustomerNoteEditor` and `showTakePaymentScreen`) would ideally go through the viewModel, so we can have unit tests for it. Same thing applies to `validateEmail`.

I think the benefits of strictly separating the view from the business logic are not so visible on as simple fragment such as this one. However, the more logic the screen contains the benefits become apparent - it's hard to guess now if this screen will become complex or not. Understanding complex screens can take days/weeks of investigation (consider for example OrderDetail in Woo or EditPostActivity in WP). 

When they are unit tested the developer doesn't need to understand most parts of the code. They can make the change they need and rely that as long as unit tests are passing the change didn't have unintended side-effects.

Moreover, when the developer reads the names of the tests, they usually get a pretty good idea what the screen is about and how it changes depending on different conditions. I sometime open unit tests for a viewmodel/screen I don't know at all and read the test names - if the tests are well written, it usually gives me a very good idea of how the code in the viewmodel works and what it does before I even saw it - eg try reading test names of [CardReadyPaymentViewModel](https://github.com/woocommerce/woocommerce-android/blob/345a285e0cacb46269d8fd2e98a2cfa8813a2da6/WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModelTest.kt#L130) - it's a pretty complex screen, but the unit test names give you a pretty good idea what's happening there.

As you can see I'm very passioned about the topic. Moreover, I think unit tests are especially beneficial when most of the business logic is covered - when some viewmodels are covered just partially, one of the main benefits is gone -> we can't change the code and rely on the unit tests to verify we didn't break anything as the thing we broke might be in the not-unit-tested part.

I'll be happy to discuss this more and provide more examples. We could also consider doing a group-programming and refactor one of the screens into a testable and tested state. Just let me know. Sorry, for the long message on an open PR, but as mentioned at the beginning I don't suggest making any changes in this PR.

**Review by malinajirka — APPROVED** [review]
Thanks for the changes @nbradbury! I especially like that you didn't only update the comments/tests but you iterated on it and improved names/comments in follow up commits ❤️ !

It works as described! I'm merging this PR to unblock progress. The one thing I'd consider is updating "Taxes are automatically calculated based on your store address" text when "automatic calculation" is disabled on the store, it feels a bit misleading.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragment.kt (hunk 2 above)
Comment (nbradbury) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragment.kt:104]: @malinajirka First off, thanks for the review!

In general I agree with you and I understand the purpose and necessity of tests. What I'm cautious about is how treating tests as a silver bullet can lead to added complexity, and complexity can lead to added problems. 

We have so much indirection in our code base in order to make it testable. For example, the user clicks a button in a fragment, the fragment tells the `ViewModel` the button was clicked, the `ViewModel` changes its `viewState` which is observed by the fragment, and the fragment responds to that observed change. All of this is done to make the code testable, but IMO it also makes the code harder to read and harder to debug.

This is not to say I'm "opposed" to tests - quite the opposite, I get their importance. I'm simply opposed to added complexity in the name of testing, and I continue to hope a better approach to tests will surface.

Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/SimplePaymentsFragment.kt:104]: > We have so much indirection in our code base in order to make it testable. For example, the user clicks a button in a fragment, the fragment tells the ViewModel the button was clicked, the ViewModel changes its viewState which is observed by the fragment, and the fragment responds to that observed change. All of this is done to make the code testable, but IMO it also makes the code harder to read and harder to debug.

@nbradbury I see your point, but I look at it from a slightly different point of view. For me what makes code more difficult to follow and write is when classes don't have clear responsibilities. When I'm writing code and the responsibilities are not defined I spent unnecessary time thinking where to put things. When I'm reading code and responsibilities are not defined, I don't know which classes I'm supposed to check to see what might cause a certain issue. Moreover, it's very easy to miss that a screen does something unless I check all the layers very carefully. Not mentioning the extra effort required when I want to add a new feature to/update the code.

For me the view layer has only three responsibilities -> propagate user events to the view model, map observed view state to UI and navigation (ideally navigation, wouldn't be there, but that's another topic). The view model (layer) is responsible for creating the view state and updating it based on user and system events.

Putting all the business logic into the ViewModel isn't only about testing. It's mostly about having clear layering which in turn makes reading/writing the code easier, testing possible and future improvements faster. Even though I agree that this layering adds some indirection+boilerplate (takes a bit of extra time), it imo and in my experience pays off extremely quickly - eg. I don't need to run the app to see if the screen is working, I can just run unit tests. Of course I always run the app just to be 100% sure I didn't miss any mappings, but running unit tests is so much faster then running the app and I usually need to run the app just once. Not mentioning all the other benefits such as "refactoring is easier", "all future additions are easier", "test serve as documentation of the intended behavior", .....

To summarize, I agree that for example the navigation requires a bit of extra boilerplate and indirection. However, if we follow the layering everywhere, all the code becomes imo easier to follow - because the added indirection and boiler plate is identical no matter the screen. If we use this indirection+boilerplate on one screen but not the other, it just makes everything more complex imo. I also believe that the benefits of having the code layered incomparably overweights the downsides of the extra indirection+boilerplate.
